### PR TITLE
Add a mail manager and send notification emails to followers of new deal poster

### PR DESCRIPTION
**Pull request by jyx11011**
Emails are only sent out when running live server

### CODE REVIEW

**Review by limli — DISMISSED** [review]

File: src/test/java/com/google/step/datamanager/MailManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package com.google.step.datamanager;
+
+public class MailManagerTest {}
```
Comment (limli) [inline src/test/java/com/google/step/datamanager/MailManagerTest.java]: is this file supposed to be empty?

File: src/main/java/com/google/step/datamanager/MailManager.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        e.printStackTrace();
+        continue;
+      }
+    }
+  }
+
+  private String composeEmail(User recipient, Deal newDeal, User poster) {
+    return String.format(
+        "Dear %s,\n"
+            + "There is a new deal posted by %s: <a href='/deals/%s'>%s</a>.\n\n"
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/MailManager.java]: will the link work if you don't specify the full path? Like how would the email know where "/deals/123" goes to?
And is it possible that the poster's username might have some html tags? Like if the poster's username is ``

**Review by limli — DISMISSED** [review]

File: src/main/java/com/google/step/datamanager/MailManager.java (hunk 2 above)
Comment (jyx11011) [reply src/main/java/com/google/step/datamanager/MailManager.java]: Reasonable

Comment (jyx11011) [reply src/main/java/com/google/step/datamanager/MailManager.java]: I will validate the username in another PR.

**Review by limli — DISMISSED** [review]

File: src/main/java/com/google/step/datamanager/MailManager.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   * @param recipients recipients of the notification email.
+   * @param poster poster of the new deal.
+   */
+  public void sendNewPostNotificationMail(List<User> recipients, Deal newDeal, User poster) {
+    if (recipients.isEmpty()) {
+      return;
+    }
+
+    Properties props = new Properties();
+    Session session = Session.getDefaultInstance(props, null);
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/MailManager.java]:
```suggestion
    Session session = Session.getDefaultInstance(new Properties(), null);
```

File: src/main/java/com/google/step/datamanager/MailManager.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    Message msg = new MimeMessage(session);
+    try {
+      msg.setFrom(new InternetAddress(SENDER, "DealFinder Team"));
+      msg.setSubject("New deal post from " + poster.username);
+    } catch (MessagingException | UnsupportedEncodingException e) {
+      e.printStackTrace();
+      return;
+    }
+
+    for (int i = 0; i < recipients.size(); i++) {
```
Comment (SuyashLakhotia) [inline src/main/java/com/google/step/datamanager/MailManager.java]:
```suggestion
    for (User recipient : recipients) {
```

File: src/test/java/com/google/step/servlets/DealServletTest.java [hunk 5]
```diff
@@ -116,6 +137,54 @@ public void testDoPost_success() throws IOException {
...
+    ArgumentCaptor<com.google.step.model.User> posterCaptor =
+        ArgumentCaptor.forClass(com.google.step.model.User.class);
+    ArgumentCaptor<List> recipientsCaptor = ArgumentCaptor.forClass(List.class);
+    ArgumentCaptor<Deal> dealCaptor = ArgumentCaptor.forClass(Deal.class);
+    verify(mockMailManager)
+        .sendNewPostNotificationMail(
+            recipientsCaptor.capture(), dealCaptor.capture(), posterCaptor.capture());
+    assertEquals(Arrays.asList(USER_B), recipientsCaptor.getValue());
+    assertEquals(DEAL_A, dealCaptor.getValue());
+    assertEquals(USER_A, posterCaptor.getValue());
```
Comment (SuyashLakhotia) [inline src/test/java/com/google/step/servlets/DealServletTest.java]:
```suggestion
    verify(mockMailManager)
        .sendNewPostNotificationMail(Arrays.asList(USER_B), DEAL_A, USER_A);
```

Does this work?

**Review by SuyashLakhotia — CHANGES_REQUESTED** [review]

**Review by SuyashLakhotia — APPROVED** [review]

**Pull request merged**